doc/ip/next_hop_group_hld.md

```diff
@@ -0,0 +1,148 @@
+
+When the next hop group is deleted, the next hop group orchagent will remove the next hop group from the ASIC_DB if required, and remove the association between the group identifier from APP_DB and the SAI identifier from ASIC_DB.  This will only happen once the next hop group is no longer referenced by any routes - and so the next hop group will maintain a reference count to ensure this.
+
+The route orchagent will parse the new nexthop_group field, and call into the next hop group orchagent to find the corresponding SAI object ID to program as the next hop attribute.  If the next hop group does not exist, the route will remain on the list of routes to sync, and will be retried on future runs of the route orchagent.
+
+If a next hop group cannot be programmed because the data plane limit has been reached, one next hop will be picked to be temporarily used for that group.  When a route is then programmed using that next hop group, it will be informed that a temporary form of the next hop group is in use, and so the route will remain on the pending list and will retry until the correct next hop group is programmed.  This mirrors the current behaviour in the route orchagent when the next hop group limit is reached.
+
+A few notes about this design:
+
+- The next hop groups created by the next hop group orchagent will not interact with those created by the route orchagent's current mechanisms. This is because the route orchagent's groups are identified by the members of the group, and so changing those members makes it a different group, whereas the new next hop group orchagent's groups have an arbitrary index, with members able to change without it becoming a different group. Therefore even if a group in each category share the same members, they will be programmed separately into ASIC_DB, in case the members change. The expectation is that all routes will be programmed into APP_DB using the old next hop schema or the new one, and so the only occasion where next hop groups could overlap is if a next hop group is created for an ACL that overlaps with one created by the next hop group orchagent.
+- A ROUTE_TABLE entry with both next hop information embedded and a reference to the NEXT_HOP_GROUP_TABLE will be ignored.
+- Next hop groups created by the fine grained next hop group orchagent will be unchanged by this design.
+
+## 3.4 SyncD
+No change.
+
+
+## 3.5 SAI
+No changes are being made in SAI. The end result of what gets programmed over the SAI will be the same as currently gets programmed - the difference is the programming sequence to get there.
+
+## 3.6 CLI
+
+### 3.6.1 Show Commands
+
+The output of 'show ip route' and 'show ipv6 route' will remain unchanged - the CLI code will resolve the next hop group ID referenced in the ROUTE_TABLE to display the next hops for the routes.
+
+# 4 Warm Boot Support
```

Comment:
>Warm Boot Support [](start = 4, length = 17)

What abut the fast convergence feature we have developed, does it apply to this next group design?

Reply:
[https://github.com/Azure/SONiC/blob/master/doc/fast-reboot/fastreboot.pdf](https://github.com/Azure/SONiC/blob/master/doc/fast-reboot/fastreboot.pdf)

Reply:
Added a comment about this